text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+
+**Semantic Conventions -** a schema defining the attributes which describe common concepts and operations which the signal observes. Note that unlike the API or SDK, Stable conventions for all signals may be placed in the same package, as they are often useful across different signals.
+
+**Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term contrib specifically refers to the plugins and instrumentation maintained by the OpenTelemetry SIG; it does not refer to third party plugins hosted elsewhere.
+
+## Signal lifecycle
+
+OpenTelemetry is structured around signals. Each signal represents a coherent, stand-alone set of functionality. Each signal follows a lifecycle.
+
+![drawing](img/0143_api_lifecycle.png)
+
+### Lifecycle stages
+
+**Experimental –** Bugs, breaking changes and performance issues may occur.
+
+**Stable –**  Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.
+
+**Deprecated –** this signal has been replaced but is still supported under LTS.
+
+**Removed -** a deprecated signal is no longer supported, and is removed.
+
+All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.
+
+### Stability
+
+Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.
+
+**API Stability -**
+No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.
+
+**SDK Stability -**
+Public portions of the SDK (constructors, configuration, und-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.
+
+**Semantic Conventions Stability -**
+Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed.
+
+**Contrib Stability -**
+Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.
+
+Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility not required.
+
+Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard.
```

Comment:
I'm not sure what this ends up implying. What does it mean in practice? If a bug is found in some instrumentation, and the results are different telemetry, does that instrumentation have to have a major version bump?

Reply:
@jkwatson Interesting point. I think bug fixing should be the only valid reason to 'break' this invariant.

Reply:
@jkwatson that sounds reasonable to me. Contrib packages should have their own version numbers, and if they break existing data, then it's a major version bump for that package. But I'm not sure how painful this is to manage in practice. 

Maybe there is some practical advice to be learned from New Relic, Dynatrace, etc, about managing the balance the between preserving existing data and making changes and improvements to instrumentation.

Reply:
I'm just thinking of the opentelemetry-java-instrumentation agent. If every change in telemetry led to a new major version of the agent, I think that might get quite unwieldy.  @trask opinions?

Reply:
@jkwatson ya it's tricky, let's discuss with @iNikem and @anuraaga in [tomorrow night's SIG meeting](https://calendar.google.com/event?action=TEMPLATE&tmeid=Nm1vZXBtNmxxMm44dW9sOWRmZWwwdm5icnRfMjAyMDEyMDlUMDYwMDAwWiBnb29nbGUuY29tX2I3OWUzZTkwajdiYnNhMm4ycDVhbjVsZjYwQGc&tmsrc=google.com_b79e3e90j7bbsa2n2p5an5lf60%40group.calendar.google.com)

@tedsuo if it's not too late, would be great to have you join

I'll post back here on Wed with thoughts from the team

Reply:
Do we have a good handle on what "compatible telemetry" means?   I believe the definition is actually vendor / backend-specific.   

E.g. "Is it acceptable to add a new label to a metric produced by a contrib module"?  The strictest definition of "compatible" would be 'never', but *likely* we can be more lenient here and provide a path for this telemetry to slowly evolve to account for new use cases or issues with previous versions.

Reply:
> The strictest definition of "compatible" would be 'never'

That's not actually true. "Compatible" is defined by Semver.

Reply:
But if breaking changes depend on the backend, the instrumentation authors may not even know what is breaking for one vendor/backend vs. another.

Reply:
Dang, sorry @trask , I was on the road and just missed the meeting.

I was not considering that the agent bundles up instrumentation in Java. I think whatever you decide as a group will be more informed than my pot shots, but it does seem more useful for the version of the agent to be separate from the version of the plugins. But I can see how decoupling the two could really increase maintenance costs. 

If users have to update to the latest plugin to get the latest agent, then breaking instrumentation data should probably bump the major version number. But again, I think the Java crew can thread the needle on this better than I can.

@jsuereth I suggest that versioning for instrumentation data should just be like semver and the rest of OTel. Adding more data is backwards compatible, removal and mutation of existing data is not. 

@jkwatson In the case of a bug, fixing it would be a patch, unless the "bug" we're talking about is some kind of useful data which users rely upon. But that sounds like something other than a bug. I think some subjective reasoning would be needed in each case, to decide on whether anyone would realistically be dependent on the bug. The other option is to be pedantic, but I'm not sure if that would better serve anyone's needs in this case.

Reply:
@tedsuo Regarding instrumentation, I guess i'm trying to understand better.   Here's some scenarios:

1. I add a new label to a metric in a importer/collector, by this definition it's not "incompatible"
2. On some backends, this new label may break existing dashboards, queries, alerts or other usage of that metric.  This would be unexpected if I thought the version bump was compatible.

What I'm asking for is a *clear* definition of what "backwards compatible" means for metrics.   Should all interaction with that metric in a downstream system be backwards compatible?  I think we'll need to define what compatible metric changes are (in OpenTelemetry's definition) so vendors can adopt their own adaptations to this if their backends don't have similar compatibility.

I may split this discussion into a separate thread with some more thoughts.

Reply:
the plan from java-instrumentation is to keep all artifacts (plugins, javaagent) at the same version, massively deprecate if we have to, but not introduce any breaking changes until OTel Java API goes to v2, at which point the java-instrumentation repo would go to v2 as well.

for experimental (non-GA) instrumentation artifacts, the preference is to use version suffix, e.g. `1.15.0-alpha` for any experimental artifacts released as part of `1.15.0` release train.

Reply:
@jsuereth I don't see how adding a new label would break a dashboard? 

If you have a more nuanced proposal I'm happy to discuss it. But for now, I am going to stick to "removal and mutation of existing data" as the clear definition of backwards compatibility. Users can expect that additional spans, metrics, attributes, and events may be added and take that into account when upgrading.

Reply:
@tedsuo This may be a GCP specific thing, but adding a label where one did not exist previously can alter how timeseries are returned, depending on what aggregation you're performing in the query.   

IF the label-set is different and you write a query/alert per-label-set, you can window up with a scenario where your timeseries forks from being *one* set of points to N (based on the cardinality of the new label), which could lead to alerts.  This could affect the threshold of an alert operating on raw timeseries, or not aggregated across that label.

From an OpenTelemetry perspective, no data is lost.  From a user perspective, they just upgraded a library and suddenly  they have to adjust their queries to make sure they're aggregating this new label and not operating on raw timeseries.

Reply:
^ cc @jmacd